rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1059 @@
+    ''
+      echo hi
+    ''
+    test
+
+# We can use a more compact form as long as all inbetween lines are indented
+bar // {
+  x = 10;
+  y = 20;
+} // baz
+
+# Bad, we can't use the more compact form because an intermediate line is not indented.
+{
+  x = 10;
+  y = 20;
+} // foo // {
+  z = 30;
+  w = 40;
+}
+```
+
+### if
+
+**Desciption**
+
+- `if` and `else` keywords always start a line, the if and else bodies are indented.
+- If the condition does not fit onto one line, then it will start on the next line with indentation, and `then` will be on the start of the line following the condition.
+- `else if` chains are treated as one long sequence, with no indentation creep on each step.
+- Only simple `if` statement can be single-line, no `else if` chains.
+
+**Examples**
+
+```nix
+#1
+if builtins.length matches != 0 then
+  { inherit path matches; }
+else if path == /. then
+  [
+    1
+    2
+  ]
+else
+  go (dirOf path);
+
+#2
+if
+  matches != null
+  && builtins.length matches != 0
+then
+  { inherit path matches; }
+else if path == /. then
+  null
+else
+  go (dirOf path);
+```
+
+**Rationale and alternatives**
+
+- Attribute sets and lists could start on the same line as the if keywords, saving an indentation level on their body:
+  ```nix  
+  #1a
+  if builtins.length matches != 0 then {
+    inherit path matches;
+  } else if path == /. then [
+    1
+    2
+  ] else
+    go (dirOf path);
+  ```
+  - This results in inconsistent vertical start of the keywords, making the structure harder to follow
+- Have the `then` on the start of the next line, directly followed by the if body:
+  ```nix
+  #1b
+  if builtins.length matches != 0
+    then { inherit path matches; }
+  else if path == /.
+    then [
+      1
+      2
+    ]
+  else go (dirOf path);
+
+  #1c
+  if builtins.length matches != 0
+  then { inherit path matches; }
+  else if path == /.
+  then [
+    1
+    2
+  ]
+  else go (dirOf path);
+  ```
+
+### with, assert
+
+**Description**
+
+- The body after the statement starts on a new line, without indentation.
+- There may be exceptions for common idioms, in which the body already starts on the same line.
+
+**Examples**
+
+```nix
+with pkgs;
+assert foo == bar; {
+  meta.maintainers =
+    with lib.maintainers; [
+      some
+      people
+    ];
+}
+```
+
+### let
+
+**Description**
+
+- Let bindings are *always* multiline.
+- The "let" part is indented one level, but not the "in" part.
+- Each item in the "let" part is indented and starts on its own line.
+  For more details, see the [binders section](#binders).
+- The "in" part starts on a new line.
+
+**Examples**
+
+```nix
+let
+  foo = "bar";
+in
+if foo == "bar" then
+  "hello"
+else
+  "world"
+```
+
+### Attribute sets and lists
+
+**Description**
+
+- Brackets and braces are generally written with a space on the inside, like `[ `, ` ]`, `{ ` and ` }`.
+  - Empty lists and attribute sets are written as`[ ]` and`{ }`, respectively.
```

Comment:
```suggestion
  - Empty lists and attribute sets are written as `[]` and`{}`, respectively.
```

This is a personal plea. Formatting rules should be practical even without tooling. On QWERTY without brace completion, and assuming regular typing ability, it's really annoying to type `Shift+[`, `Space`, `Shift+]` to express `{ }`. Without the space, you hold `Shift` and roll your fingers over `[` and `]`. Somewhat minor: If you have the formatter running automatically it will let the code jump needlessly on this trivial thing.

Arguably we should optimize for the still average case of QWERTY keyboards. Everyone else will know what they're doing anyway.

Reply:
nit: Technically for the vanilla vim keybinds you can do `s`, `Enter`, `Tab`.
But I fully agree with you, and never leave a space in there as I don't really see the point.

Reply:
> Technically for the vanilla vim keybinds you can do `s`, `Enter`, `Tab`.

TIL

Reply:
This is something I really don't think matters a lot. We concluded that `{ }` is more consistent with `{ foo = 10; }`, so we chose that. I don't think anybody would be annoyed by this in practice.

Reply:
> On QWERTY without brace completion, and assuming regular typing ability, it's really annoying to type `Shift+[`, `Space`, `Shift+]` to express `{ }`. Without the space, you hold `Shift` and roll your fingers over `[` and `]`.

Just hold shift the entire time. `Shift + Space` enters a space char. Only weird layouts like Neo differentiate here.

> Even more minor: When expanding a list or attrset with Vim-style controls, navigating to the space and hitting `r`, `Enter`, `i`, `Tab` is one interaction more than `w` to the first brace `a`,`Enter`,`Tab`.

In properly configured editors, you'd simply enter insert mode somewhere between the braces and press enter. It should automatically remove possible superfluous spaces at the end of the line and indent on enter.

This is an editor configuration problem, not a formatting one IMHO.

Reply:
> In properly configured editors

If you're targeting software professionals, that's what you can assume. If you want to be inclusive to a wide range of people from different backgrounds (and we've seen some reports where the Nix language was the first encounter with programming), you have to take into account that it's not a given.

Again, not fighting here, just raising considerations I'd like people participating in this discussion to be aware of. RFC authors feel free to resolve close this thread at will.

Reply:
While I don't find the ease of typing argument strong (though, it's likely that I'm biased as a "software professional"), I do think that requiring a space for empty attrsets and list looks strange, especially if you consider that probably most Nix users will have background in other programming languages. In my experience with various mainstream languages, formatters prefer `[]` and `{}` over `[ ]` and `{ }`.

I did some quick testing online and C# was the only exception:
* `dotnet format` (C#) - prefers `{ }` - [link](https://replit.com/@PetarKirov/csharp-formatting-empty-objects#main.cs)

other language formatters:

* Black (Python) - prefers `[]` and `{}` - [link](https://black.vercel.app/?version=stable&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4ACOAFJdAD2IimZxl1N_WlPS8NIZ-x4TNI2_ZJeJE5zxHZZWqBkdWPX0qG2H3hsDb2YmAGRmmWBxViohAyw5oXP2m5Fm6G0jZff5LMsRaXvuQ4O38I8uXQAAAAAzOT-uPfRThAABbo8BAAAA0fqherHEZ_sCAAAAAARZWg==)
* Prettier (JavaScript, TypeScript) - prefers `[]` and `{}` - [link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBDATrzAXkwG1NyBdcgbgB0pIoNMIAjAKyM2HMwF8aIADQgIABxgBLdMlB5cEAO4AFPAjTIQ2ADaLsATw0jWubGADWcGAGVsAWzgAZSVDjIAZjrRxjpi1esxMxcAc2QYXABXHxA4O1Y4ABNEpMdsKBDI7BC4ADEIXDtsGCkMzWxImAhhEAALGDttAHVayXg0ILA4a3U2yQA3Nv1NMDQjEBdvXBhlUxCijy8Y9jQAD2tQ7TgARUiIeEXtbxEg3CnNVmwE7RqxXBcYJslEmFrkAA4ABhOFbybTMSaO5wKb9NwiACOe3gs3EGhQ2DQAFpXEkkjVcHAoZJMbNsgskJ4jjFvHZJOEoiTNnAAIIle6sSpwZRwXDOVyHY4gNDU3b7NyEpYiGBXJ4vN5IABMwtMkm0oQAwhA7ATYmgAKw1SLeAAqV3hRK5-WiAEkoClYNYwPcJDTzdYYPotpy4Hw+EA)
* Clang Format (C/C++) - prefers `{}` - tested with [this code](https://gist.github.com/PetarKirov/978c7d5ec2ca5338c762ec1cb4ec254d) at https://clang-format-configurator.site/
* `google-java-format` (Java) - prefers and `{}` - https://codebeautify.org/javaviewer/y23b1bfbb
* `dart format` (Dart) - prefers `[]` and `{}` - https://dartpad.dev/?id=50af8512ea53ee32ec05e995ebb58983
* `go fmt` (Go) - prefers `{}` - https://go.dev/play/p/9m0Zq_oz1ZO
* Ormolu (Haskell) - prefers `[]` - (tested [this code](https://gist.github.com/PetarKirov/924b09cb75eb2b87089512171501df54) on https://ormolu-live.tweag.io/)
* Rustfmt (Rust) - prefers `[]` and `{}` - [link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dfbd5ebcc73c37c827cc271ebba03d49)
* Swift  - prefers `()`, `{}`, `[]`, and `[:]` - [link](https://swiftfiddle.com/cfsqjnuayzh55fxxi6ygx2hh7y)

Reply:
I personally don't have any overly strong opinions on this either way. However, for me consistency is important here. If we write empty lists without a space, then we should also write non-empty lists without: `[foo bar]`. Same for attribute sets. (This would be more consistent with function calls too btw.)

The space version was initially chosen because in the absence of strong arguments, current formatting conventions are followed.

Reply:
In the meeting today, the shepherds agreed to consider the style without spaces, we'll post a Nixpkgs diff soon!

Reply:
I'm also for no spaces, but probably only because I am used to it through alejandra.

Reply:
I'm not a fan of no space. Lists in Nix use whitespace to separate items; either space chars or newlines. Whitespace therefore has great importance in these contexts.

```Nix
[a b c]
```

is inconsistent with

```Nix
[
  a
  b
  c
]
```

because there's suddenly whitespace between `[` and `a` where there was none before.

Additionally, I don't agree with the argument that other languages do no spaces as almost all "other languages" have "physical" separators between list items (usually `,`).

With `[a, b, c]`, it's a lot more bearable to have the `[` so close to the `a` than it is with `[a b c]` because "physical" characters in close proximity to the items is the norm.

***

As for `[]` vs `[ ]`, I feel like the whitespace further solidifies the emptiness of the list. `[]` reminds me too much of C arrays where the length is anything but clear.

Empty sets should be consistent with the empty lists and vice versa. I think we can all agree on that.

Let's look at how sets work in Nix then. In sets, we always put whitespace before/after the squirly braces like this `{ foo = "bar"; }` or `{ ... }`. This is consistent with multi-line sets where there is also always a whitespace (newline) before/after squirlies:

```Nix
{
  foo = "bar";
}
```

I'd find it extremely weird and inconsistent to suddenly not put whitespace before/after squirlies when it's a single line (`{foo = "bar";}`).

I've yet to see a good reason for omitting these spaces, especially given that almost all Nix code in existence has them.

Reply:
Tentatively implemented in https://github.com/serokell/nixfmt/pull/118/commits/c4b930210270277a658183a034e4a97374daacf0

Reply:
I was mostly agnostic, but after seeing the examples in that implementation commit I'm now firmly on team spaces.

The double-winky-grinch emoticon at the end of this line is what did it for me:
```
  {a = with b; {c = 1;};}
```

Reply:
Discussed again in today's meeting. For consistency reasons we really think that there should always be a space for now, both for attribute sets, lists, and empty versions of them. We'd like to resolve this topic as we don't think it's worth discussing further for now.